### DISCUSSION TITLE
Avoiding ECS Cold Start for EcsRunLauncher

### DISCUSSION THREAD

**Discussion by Jma353** [dagster-io/dagster#7332] [Q&A] [answered]
Hey folks, I'm looking for some guidance on the following use-case!

## Context

My team is really enjoying Dagster and the `EcsRunLauncher` is very convenient for dynamically spinning up ephemeral ECS tasks to perform work. We are getting value out of this today!

However, sometimes we want a job to finish as fast as possible (assuming no queuing) and want the job to execute *without the cold-start overhead of spinning up an ephemeral ECS task.* Our ideal outcome is that some jobs start immediately, no cold-start overhead, if flagged with the appropriate "priority".

## Questions
I have a couple of questions about this, and would love support on a Dagster-vetted solution on how we can approach this: 
1. We were thinking about writing our own launcher, which can choose how to run the job (either ECS or on a dedicated host). We were planning on allocating a set of worker hosts that were not ephemeral and able to be dispatched to if a job request came in with the right "priority". *What do you think about this?*
2. Is there a way with the current out-of-the-box Dagster tools to accomplish the above behavior described in the "Context" section? 
3. Is there work on the roadmap involving enabling folks who you use ECS as the primary deployment mechanism more?
4. How does Dagster Cloud deal with this problem, if at all? Are all jobs subject to a cold-start time?

**Comment by johannkm** [comment, marked as answer]
Hi @Jma353, yeah this is something we're thinking about internally. Currently we do pay that cold-start price everywhere that we use ECS, but it's not great. We're not far off from being able to support what you describe in (1)- the `DefaultRunLauncher` sends runs to the standing gRPC server. Dagster doesn't currently support swapping run launchers at runtime, but until then it probably wouldn't be too hard to write a custom run launcher that wraps both run launcher options, and uses one depending on a run tag or something like that

**Reply by george-chamoun** [reply]
Thanks for this tip. I stumbled upon this thread with a similar question about swapping run launchers (although for slightly different reasons). I will go down this route, as I think use cases will vary too much by user for someone to ask you to provide a generic solution. For example, if all jobs are small enough to run on the gRPC server / the size is not very variable, sure, sending to the gRPC until the ECS task is live a fine solution. However, this will not work for everyone. My runs vary greatly in size. Therefore, I will need custom logic to determine which jobs are worth waiting for the ECS start and which are small enough to just send to the gRPC server. Thanks again.

**Reply by vjindal0112** [reply]
Hi, curious if there's any update here? The cold start time is prohibitive for our team to use Dagster for more use cases.

**Reply by johannkm** [reply]
Hi Varun:

- We prioritized getting this fixed in Dagster Cloud, and did so by partially giving up on Docker containers: https://dagster.io/blog/fast-deploys-with-pex-and-docker. Currently this is only available in the Serverless offering, but we'll be adding it to Hybrid.
- We also added 'non-isolated runs' to Cloud, which are launched as processes on the standing gRPC server rather than in new tasks. This is basically the approach that users can take by implmementing the run launcher switching between the EcsRunlauncher and the DefaultRunLauncher. We've run in to some issues with this approach- e.g. it's easy to OOM the grpc server, and hard to communicate why that happened since it can be running multiple sensors and non isolated runs at once. As a result we haven't added native support for this in OSS, but again it's achievable with a custom run launcher. We'd just recommend throwing more resources at the gRPC server.
- In general I believe the problem is reduced if you run ECS on EC2, where you can manually plan capacity and get image caching. Versus Fargate which is slow.

Happy to talk more about your use cases here or in our Slack.

**Reply by jzgrzebnicki** [reply]
Hi, to speed up things we move from fargate to Ec2 but not without problems. We have two ec2, but EcsRunLauncher spins up taskdef in VPC mode and some ec2 have limit on ENI, so we needed to create our own task definition in a BRIDGE mode, now our runs start much faster.

**Reply by skeller88** [reply]
@jzgrzebnicki Did you write your own EC2 RunLauncher?

**Reply by jzgrzebnicki** [reply]
No we defined our own task definition in Bridge, so there was no need to write it.
